File: utils/handlers/quests.js
const questService = require('../../services/questService');
const questPool = require('../data/quests');
const ui = require('../ui');
const { Markup } = require('telegraf');

/**
 * Quest Handler: UI and interaction for the daily quest system.
 */
module.exports = {
    async handleQuests(ctx) {
        if (ctx.callbackQuery) await ctx.answerCbQuery().catch(() => null);
        const userId = ctx.from.id;
        const quests = await questService.syncQuests(userId);
        const resetTime = questService.getTimeUntilReset();

        let msg = ui.formatHeader("📋 DAILY QUESTS") + "\n" +
            `⏳ Resets in: <b>${resetTime}</b>\n` +
            `${ui.divider()}\n\n`;

        const kb = [];

        quests.forEach((uq, idx) => {
            const meta = questPool[uq.questId];
            const statusIcon = uq.claimed ? "✅" : (uq.completed ? "🌟" : "⬜");
            const progressPct = Math.min(1, uq.progress / uq.target);
            const progressBar = this.generateProgressBar(progressPct);
            
            msg += `${statusIcon} <b>${idx + 1}. ${meta.description}</b>\n` +
                `   ${progressBar} (${uq.progress}/${uq.target})\n` +
                `   💰 <i>Rewards: ${meta.reward.coins} coins + ${meta.reward.xp} XP</i>\n\n`;

            if (uq.completed && !uq.claimed) {
                kb.push([Markup.button.callback(`🎁 CLAIM QUEST ${idx + 1}`, `q_claim_${uq.questId}`)]);
            }
        });

        msg += `${ui.divider()}`;
        
        const hasUnclaimed = quests.some(q => q.completed && !q.claimed);
        if (hasUnclaimed) {
            kb.push([Markup.button.callback('✨ CLAIM ALL REWARDS', 'q_claim_all')]);
        }
        
        kb.push([Markup.button.callback('🔙 BACK', 'back_to_hub')]);

        if (ctx.callbackQuery) {
            return ctx.editMessageText(msg, { parse_mode: 'HTML', ...Markup.inlineKeyboard(kb) }).catch(() => null);
        }
        return ctx.replyWithHTML(msg, Markup.inlineKeyboard(kb));
    },

    generateProgressBar(pct) {
        const size = 10;
        const filled = Math.round(size * pct);
        const empty = size - filled;
        return `[${'█'.repeat(filled)}${'░'.repeat(empty)}]`;
    },

    async handleClaim(ctx, questId) {
        await ctx.answerCbQuery("Processing claim...").catch(() => null);
        const result = await questService.claimQuest(ctx.from.id, questId);
        if (result.success) {
            let rewardMsg = "✅ <b>Quest completed!</b> You received:\n\n" +
                `💰 <b>Coins:</b> +${result.reward.coins}\n` +
                `📈 <b>Experience:</b> +${result.reward.xp}`;
            
            if (result.reward.items) {
                result.reward.items.forEach(i => rewardMsg += `\n📦 <b>${i.id}:</b> +${i.qty}`);
            }

            await ctx.replyWithHTML(rewardMsg);
            return this.handleQuests(ctx);
        } else {
            return ctx.replyWithHTML(`❌ <b>Claim failed:</b> ${result.message}`);
        }
    },

    async handleClaimAll(ctx) {
        await ctx.answerCbQuery("Focusing energies...").catch(() => null);
        const result = await questService.claimAll(ctx.from.id);
        if (result.success) {
            let rewardMsg = "✨ <b>All quests claimed!</b> Total rewards:\n\n" +
                `💰 <b>Coins:</b> +${result.coins}\n` +
                `📈 <b>Experience:</b> +${result.xp}`;
            
            if (result.items && result.items.length > 0) {
                // Group items
                const consolidated = {};
                result.items.forEach(i => consolidated[i.id] = (consolidated[i.id] || 0) + i.qty);
                Object.entries(consolidated).forEach(([id, qty]) => {
                    rewardMsg += `\n📦 <b>${id}:</b> +${qty}`;
                });
            }

            await ctx.replyWithHTML(rewardMsg);
            return this.handleQuests(ctx);
        } else {
            return ctx.replyWithHTML(`❌ <b>Claim failed:</b> ${result.message}`);
        }
    }
};

File: utils/handlers/quests.py
from aiogram import Router, types, F
from aiogram.utils.keyboard import InlineKeyboardBuilder
from aiogram.filters import Command

from utils import ui
from services.quest_service import quest_service
from utils.data import quests as quest_pool

router = Router()

@router.message(Command("quests"))
@router.callback_query(F.data.startswith("cmd_quests"))
async def handle_quests(callback_or_message: types.CallbackQuery | types.Message, user: dict):
    if isinstance(callback_or_message, types.CallbackQuery):
        await callback_or_message.answer()
    
    user_id = callback_or_message.from_user.id
    active_quests = await quest_service.sync_quests(user_id)
    reset_time = quest_service.get_time_until_reset()

    msg = ui.format_header("📋 DAILY QUESTS") + "\n" + \
          f"⏳ Resets in: <b>{reset_time}</b>\n" + \
          f"{ui.divider()}\n\n"

    builder = InlineKeyboardBuilder()

    for idx, uq in enumerate(active_quests):
        meta = quest_pool.DATA.get(uq['questId'])
        if not meta: continue
        
        status_icon = "✅" if uq.get('claimed') else ("🌟" if uq.get('completed') else "⬜")
        progress_pct = min(1.0, uq['progress'] / uq['target'])
        
        # UI Progress Bar
        size = 10
        filled = round(size * progress_pct)
        empty = size - filled
        progress_bar = f"[{'█' * filled}{'░' * empty}]"
        
        msg += f"{status_icon} <b>{idx + 1}. {meta['description']}</b>\n" + \
               f"   {progress_bar} ({uq['progress']}/{uq['target']})\n" + \
               f"   💰 <i>Rewards: {meta['reward'].get('coins', 0)} coins + {meta['reward'].get('xp', 0)} XP</i>\n\n"

        if uq.get('completed') and not uq.get('claimed'):
            builder.row(types.InlineKeyboardButton(text=f"🎁 CLAIM QUEST {idx + 1}", callback_data=f"q_claim_{uq['questId']}"))

    msg += f"{ui.divider()}"
    
    has_unclaimed = any(q.get('completed') and not q.get('claimed') for q in active_quests)
    if has_unclaimed:
        builder.row(types.InlineKeyboardButton(text="✨ CLAIM ALL REWARDS", callback_data="q_claim_all"))
    
    builder.row(types.InlineKeyboardButton(text="🔙 BACK", callback_data="back_to_hub"))

    if isinstance(callback_or_message, types.CallbackQuery):
        await callback_or_message.message.edit_text(msg, parse_mode='HTML', reply_markup=builder.as_markup())
    else:
        await callback_or_message.answer(msg, parse_mode='HTML', reply_markup=builder.as_markup())

@router.callback_query(F.data.startswith("q_claim_"))
async def handle_claim(callback: types.CallbackQuery):
    if callback.data == "q_claim_all":
        await callback.answer("Focusing energies...")
        result = await quest_service.claim_all(callback.from_user.id)
        if result['success']:
            msg = f"✨ <b>All quests claimed!</b> Total rewards:\n\n💰 <b>Coins:</b> +{result['coins']}\n📈 <b>Experience:</b> +{result['xp']}"
            if result.get('items'):
                consolidated = {}
                for i in result['items']: consolidated[i['id']] = consolidated.get(i['id'], 0) + i['qty']
                for tid, qty in consolidated.items(): msg += f"\n📦 <b>{tid}:</b> +{qty}"
            await callback.message.reply(msg, parse_mode='HTML')
            await handle_quests(callback)
        else:
            await callback.answer(result['message'], show_alert=True)
    else:
        quest_id = callback.data.replace("q_claim_", "")
        await callback.answer("Processing claim...")
        result = await quest_service.claim_quest(callback.from_user.id, quest_id)
        if result['success']:
            msg = f"✅ <b>Quest completed!</b> You received:\n\n💰 <b>Coins:</b> +{result['reward']['coins']}\n📈 <b>Experience:</b> +{result['reward']['xp']}"
            if result['reward'].get('items'):
                for i in result['reward']['items']: msg += f"\n📦 <b>{i['id']}:</b> +{i['qty']}"
            await callback.message.reply(msg, parse_mode='HTML')
            await handle_quests(callback)
        else:
            await callback.answer(result['message'], show_alert=True)
